File: 백준/Gold/9251. LCS/LCS.py
# 백준 9251

def lcs(s1, s2):
    len1, len2 = len(s1), len(s2)
    arr = [[0]*(len2+1) for _ in range(len1+1)]
    for i in range(1, len1+1):
        for j in range(1, len2+1):
            if s1[i-1]==s2[j-1]:                # 비교하는 문자가 같다면
                arr[i][j] = arr[i-1][j-1]+1
            else:
                arr[i][j] = max(arr[i][j-1], arr[i-1][j])

    return arr[-1][-1]

s1 = input()
s2 = input()

print(lcs(s1, s2))

File: 백준/Gold/9251. LCS/LCS.java
import java.io.*;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String s1 = br.readLine();
        String s2 = br.readLine();
        int answer = lcs(s1, s2);
        System.out.println(answer);
    }

    public static int lcs(String s1, String s2) {
        int len1 = s1.length();
        int len2 = s2.length();

        int[][] arr = new int[len1+1][len2+1];


        for (int i = 1; i<len1+1; i++) {
            for (int j = 1; j<len2+1; j++) {
                if (s1.charAt(i-1) == s2.charAt(j-1)) {
                    arr[i][j] = arr[i-1][j-1] + 1;
                } else {
                    arr[i][j] = Math.max(arr[i-1][j], arr[i][j-1]);
                }
            }
        }
        return arr[len1][len2];
    }
}